libpyvinyl/Parameters/Parameter.py

```diff
@@ -0,0 +1,181 @@
+# Created by Mads Bertelsen and modified by Juncheng E
+
+import math
+from libpyvinyl.AbstractBaseClass import AbstractBaseClass
+
+
+class Parameter(AbstractBaseClass):
+    """
+    Description of a single parameter
+    """
+    def __init__(self, name, unit=None, comment=None):
+        """
+        Creates parameter with given name, optionally unit and comment
+        """
+        self.name = name
+        self.unit = unit
+        self.comment = comment
+        self.value = None
+        self.legal_intervals = []
+        self.illegal_intervals = []
+        self.options = []
+
+    @classmethod
+    def from_dict(cls, param_dict):
+        param = cls(param_dict['name'], param_dict['unit'],
+                    param_dict['comment'])
+        for key in param_dict:
+            param.__dict__[key] = param_dict[key]
+        return param
+
+    def add_legal_interval(self, min_value, max_value):
+        """
+        Sets a legal interval for this parameter, None for infinite
+        """
+        if min_value is None:
+            min_value = -math.inf
+        if max_value is None:
+            max_value = math.inf
+
+        self.legal_intervals.append([min_value, max_value])
+
+    def add_illegal_interval(self, min_value, max_value):
+        """
+        Sets an illegal interval for this parameter, None for infinite
+        """
+        if min_value is None:
+            min_value = -math.inf
+        if max_value is None:
+            max_value = math.inf
+
+        self.illegal_intervals.append([min_value, max_value])
+
+    def add_option(self, option):
+        """
+        Sets allowed values for this parameter
+        """
+        if isinstance(option, list):
+            self.options += option
+        else:
+            self.options.append(option)
+
+    def set_value(self, value):
+        """
+        Sets value of this parameter if value is legal, otherwise warning is shown
+
+        This could be expanded to raise an exception, or such could be in is_legal
+        """
+        if self.is_legal(value):
+            self.value = value
+        else:
+            print("WARNING: Value of parameter '" + self.name
+                  + "' illegal, ignored.")
+
+    def is_legal(self, value=None):
+        """
+        Checks whether or not given or contained value is legal given constraints
+        """
+        if value is None:
+            value = self.value
+
+        for illegal_interval in self.illegal_intervals:
+            if illegal_interval[0] < value < illegal_interval[1]:
+                return False
+
+        is_outside_a_legal_interval = False
```

Comment:
I guess this has been addressed?

Reply:
To be honest I found it confusing and I spent some time trying to make all the possible cases to make sure to understand it. 
```
  is_inside_a_legal_interval = False
  for legal_interval in self.legal_intervals:
            if legal_interval[0] < value < legal_interval[1]:
                is_inside_a_legal_interval = True

        if not is_inside_a_legal_interval:
            return False
```

This form seems to me clearer.
Do as you like it most :-)

Reply:
In the piece of code Shervin suggested I believe *is_legal* will return false if there are no legal intervals defined. This can be fixed by adding "and len(legal_intervals) > 0" to the if statement with the return.

Reply:
@mads-bertelsen @shervin86 Could you please check if this implementation is correct?
https://github.com/PaNOSC-ViNYL/libpyvinyl/blob/a3a2599629712b7f6447913911733c8b50edd32b/libpyvinyl/Parameters/Parameter.py#L85-L97

Reply:
If you continue to ask, I'm obliged to continue to propose changes :-)
```
 # if there are no legal intervals defined, then the defined value is considered legal
 # otherwise it has to be checked
 is_inside_a_legal_interval = (len(self.legal_intervals) == 0)

 for legal_interval in self.legal_intervals: 
     if legal_interval[0] < value < legal_interval[1]: 
         is_inside_a_legal_interval = True 
   
 if not is_inside_a_legal_interval: 
     return False 
```